Include conditions in nodes data source

### Description

I would like to get the node conditions inside the `kubernetes_nodes` data source.
Currently only `addresses`, `allocatable`, `capacity` and `node_info` are exported.

### Potential Terraform Configuration

```hcl
data "kubernetes_nodes" "example" {}

output "node_conditions" {
  value = [for node in data.kubernetes_nodes.example.nodes : node.status.conditions]
}
```

### References

none

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

> Can you share the crash log?

Sure, here is the crash log.

```
│ Error: Plugin did not respond
│
│   with data.kubernetes_resources.nodes,
│   on main.tf line 48, in data "kubernetes_resources" "nodes":
│   48: data "kubernetes_resources" "nodes" {
│
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ReadDataSource call. The
│ plugin logs may contain more details.
╵

Stack trace from the terraform-provider-kubernetes_v2.26.0_x5 plugin:

panic: runtime error: index out of range [7] with length 7

goroutine 14 [running]:
github.com/hashicorp/terraform-provider-kubernetes/manifest/payload.sliceToTFTupleValue({0xc0044df100, 0x8, 0xc004f4f620?}, {0x24c6f28?, 0xc004f0ad80}, 0x8010103?, 0xc00294fe18)
        github.com/hashicorp/terraform-provider-kubernetes/manifest/payload/to_value.go:192 +0x785
github.com/hashicorp/terraform-provider-kubernetes/manifest/payload.ToTFValue({0x1d38720?, 0xc001310300?}, {0x24c6f28, 0xc004f0ad80}, 0xc004f42730?, 0x4184e8?)
        github.com/hashicorp/terraform-provider-kubernetes/manifest/payload/to_value.go:117 +0x6c6
github.com/hashicorp/terraform-provider-kubernetes/manifest/payload.mapToTFObjectValue(0x2060520?, {0x24c6cc0?, 0xc0048f68d0}, 0xc004f42730?, 0xc00294f878)
        github.com/hashicorp/terraform-provider-kubernetes/manifest/payload/to_value.go:259 +0x308
github.com/hashicorp/terraform-provider-kubernetes/manifest/payload.ToTFValue({0x1e5e440?, 0xc004c2a840?}, {0x24c6cc0, 0xc0048f68d0}, 0x1e60300?, 0xc94b98?)
        github.com/hashicorp/terraform-provider-kubernetes/manifest/payload/to_value.go:128 +0x9aa
github.com/hashicorp/terraform-provider-kubernetes/manifest/payload.mapToTFObjectValue(0x2060520?, {0x24c6cc0?, 0xc00497a540}, 0xc0044860c8?, 0xc00289f410)
        github.com/hashicorp/terraform-provider-kubernetes/manifest/payload/to_value.go:259 +0x308
github.com/hashicorp/terraform-provider-kubernetes/manifest/payload.ToTFValue({0x1e5e440?, 0xc004c2a810?}, {0x24c6cc0, 0xc00497a540}, 0x20b2cc0?, 0xc004e54480?)
        github.com/hashicorp/terraform-provider-kubernetes/manifest/payload/to_value.go:128 +0x9aa
github.com/hashicorp/terraform-provider-kubernetes/manifest/provider.(*RawProviderServer).ReadPluralDataSource(0xc000510400, {0x24bfd48, 0xc000bf1c50}, 0xc00068e9a0)
        github.com/hashicorp/terraform-provider-kubernetes/manifest/provider/datasource.go:178 +0x15d1
github.com/hashicorp/terraform-provider-kubernetes/manifest/provider.(*RawProviderServer).ReadDataSource(0x24bfd48?, {0x24bfd48, 0xc000bf1c50}, 0xc00068e9a0)
        github.com/hashicorp/terraform-provider-kubernetes/manifest/provider/datasource.go:28 +0x85
github.com/hashicorp/terraform-plugin-mux/tf5muxserver.(*muxServer).ReadDataSource(0x24bfd80?, {0x24bfd48?, 0xc000bf1950?}, 0xc00068e9a0)
        github.com/hashicorp/terraform-plugin-mux@v0.12.0/tf5muxserver/mux_server_ReadDataSource.go:36 +0x19f
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ReadDataSource(0xc0000ce460, {0x24bfd48?, 0xc000bf0fc0?}, 0xc000b88ff0)
        github.com/hashicorp/terraform-plugin-go@v0.20.0/tfprotov5/tf5server/server.go:699 +0x3df
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadDataSource_Handler({0x20c12c0?, 0xc0000ce460}, {0x24bfd48, 0xc000bf0fc0}, 0xc0002acd80, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.20.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:539 +0x169
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000a78000, {0x24bfd48, 0xc000bf0f30}, {0x24c80a0, 0xc000b341a0}, 0xc000bed200, 0xc000767710, 0x3543dc8, 0x0)
        google.golang.org/grpc@v1.60.0/server.go:1372 +0xe03
google.golang.org/grpc.(*Server).handleStream(0xc000a78000, {0x24c80a0, 0xc000b341a0}, 0xc000bed200)
        google.golang.org/grpc@v1.60.0/server.go:1783 +0xfec
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        google.golang.org/grpc@v1.60.0/server.go:1016 +0x59
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 54
        google.golang.org/grpc@v1.60.0/server.go:1027 +0x115

Error: The terraform-provider-kubernetes_v2.26.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

--- a/.changelog/2612.txt
+++ b/.changelog/2612.txt
@@ -1,0 +1,3 @@
+```release-note:enhancement
+Added `conditions` attribute to `kubernetes_nodes` data source, which will provide detailed node health and status information
+```

--- a/docs/data-sources/nodes.md
+++ b/docs/data-sources/nodes.md
@@ -82,7 +82,7 @@ Read-Only:
 - `allocatable` (Map of String)
 - `capacity` (Map of String)
 - `node_info` (List of Object) (see [below for nested schema](#nestedobjatt--nodes--status--node_info))
-
+- `conditions` (List of Object) (see [below for nested schema](#nestedobjatt--nodes--status--conditions))
 <a id="nestedobjatt--nodes--status--addresses"></a>
 ### Nested Schema for `nodes.status.addresses`
 
@@ -108,7 +108,17 @@ Read-Only:
 - `os_image` (String)
 - `system_uuid` (String)
 
+<a id="nestedobjatt--nodes--status--conditions"></a>
+### Nested Schema for `nodes.status.conditions`
 
+Read-Only:
+
+- `type` (String)
+- `status` (String)
+- `last_heartbeat_time` (String)
+- `last_transition_time` (String)
+- `reason` (String)
+- `message` (String)
 
 
  

--- a/kubernetes/data_source_kubernetes_nodes_test.go
+++ b/kubernetes/data_source_kubernetes_nodes_test.go
@@ -36,6 +36,10 @@ func TestAccKubernetesDataSourceNodes_basic(t *testing.T) {
 		resource.TestMatchResourceAttr(dataSourceName, "nodes.0.status.0.conditions.#", oneOrMore),
 		resource.TestCheckResourceAttrSet(dataSourceName, "nodes.0.status.0.conditions.0.type"),
 		resource.TestCheckResourceAttrSet(dataSourceName, "nodes.0.status.0.conditions.0.status"),
+		resource.TestCheckResourceAttrSet(dataSourceName, "nodes.0.status.0.conditions.0.last_heartbeat_time"),
+		resource.TestCheckResourceAttrSet(dataSourceName, "nodes.0.status.0.conditions.0.last_transition_time"),
+		resource.TestCheckResourceAttrSet(dataSourceName, "nodes.0.status.0.conditions.0.reason"),
+		resource.TestCheckResourceAttrSet(dataSourceName, "nodes.0.status.0.conditions.0.message"),
 	)
 	resource.ParallelTest(t, resource.TestCase{
 		PreCheck:          func() { testAccPreCheck(t) },

--- a/kubernetes/data_source_kubernetes_nodes_test.go
+++ b/kubernetes/data_source_kubernetes_nodes_test.go
@@ -33,6 +33,9 @@ func TestAccKubernetesDataSourceNodes_basic(t *testing.T) {
 		resource.TestCheckResourceAttrWith(dataSourceName, "nodes.0.status.0.capacity.memory", checkParsableQuantity),
 		resource.TestCheckResourceAttrSet(dataSourceName, "nodes.0.status.0.node_info.0.architecture"),
 		resource.TestCheckResourceAttrSet(dataSourceName, "nodes.0.status.0.addresses.0.address"),
+		resource.TestMatchResourceAttr(dataSourceName, "nodes.0.status.0.conditions.#", oneOrMore),
+		resource.TestCheckResourceAttrSet(dataSourceName, "nodes.0.status.0.conditions.0.type"),
+		resource.TestCheckResourceAttrSet(dataSourceName, "nodes.0.status.0.conditions.0.status"),
 	)
 	resource.ParallelTest(t, resource.TestCase{
 		PreCheck:          func() { testAccPreCheck(t) },

--- a/kubernetes/schema_node_spec.go
+++ b/kubernetes/schema_node_spec.go
@@ -124,6 +124,39 @@ func nodeStatusFields() map[string]*schema.Schema {
 				},
 			},
 		},
+		"conditions": {
+			Type:        schema.TypeList,
+			Computed:    true,
+			Description: "List of conditions describing each node's health and operational status.",
+			Elem: &schema.Resource{
+				Schema: map[string]*schema.Schema{
+					"type": {
+						Type:     schema.TypeString,
+						Computed: true,
+					},
+					"status": {
+						Type:     schema.TypeString,
+						Computed: true,
+					},
+					"last_heartbeat_time": {
+						Type:     schema.TypeString,
+						Computed: true,
+					},
+					"last_transition_time": {
+						Type:     schema.TypeString,
+						Computed: true,
+					},
+					"reason": {
+						Type:     schema.TypeString,
+						Computed: true,
+					},
+					"message": {
+						Type:     schema.TypeString,
+						Computed: true,
+					},
+				},
+			},
+		},
 	}
 }
 

--- a/kubernetes/structures_node.go
+++ b/kubernetes/structures_node.go
@@ -71,12 +71,29 @@ func flattenNodeInfo(in v1.NodeSystemInfo) []interface{} {
 	return []interface{}{att}
 }
 
+func flattenNodeConditions(conditions []v1.NodeCondition) []interface{} {
+	out := make([]interface{}, len(conditions))
+	for i, condition := range conditions {
+		m := make(map[string]interface{})
+		m["type"] = condition.Type
+		m["status"] = condition.Status
+		m["last_heartbeat_time"] = condition.LastHeartbeatTime.String()
+		m["last_transition_time"] = condition.LastTransitionTime.String()
+		m["reason"] = condition.Reason
+		m["message"] = condition.Message
+		out[i] = m
+	}
+	return out
+}
+
 func flattenNodeStatus(in v1.NodeStatus) []interface{} {
 	att := make(map[string]interface{})
 	att["addresses"] = flattenAddresses(in.Addresses...)
 	att["allocatable"] = flattenResourceList(in.Allocatable)
 	att["capacity"] = flattenResourceList(in.Capacity)
 	att["node_info"] = flattenNodeInfo(in.NodeInfo)
+	att["conditions"] = flattenNodeConditions(in.Conditions)
+
 	return []interface{}{att}
 }
 

